src/storage/store/csr_node_group.cpp

```diff
@@ -117,18 +157,79 @@ NodeGroupScanResult CSRNodeGroup::scan(Transaction* transaction, TableScanState&
+        const auto offsetInGroup = currNodeOffset % StorageConstants::NODE_GROUP_SIZE;
+        if (nodeGroupScanState.header->getStartCSROffset(offsetInGroup) >
+            nodeGroupScanState.numScannedRows) {
+            nodeGroupScanState.numScannedRows =
+                nodeGroupScanState.header->getStartCSROffset(offsetInGroup);
+        }
+        KU_ASSERT(nodeGroupScanState.numScannedRows <= nodeGroupScanState.numTotalRows);
+        const auto numToScan =
+            std::min(nodeGroupScanState.numTotalRows - nodeGroupScanState.numScannedRows,
+                DEFAULT_VECTOR_CAPACITY);
+        persistentChunkGroup->scan(transaction, tableState, nodeGroupScanState,
+            nodeGroupScanState.numScannedRows, numToScan);
+        nodeGroupScanState.numCachedRows = numToScan;
+        nodeGroupScanState.numScannedRows += numToScan;
+        if (tableState.outState->getSelVector().isUnfiltered()) {
+            nodeGroupScanState.cachedScannedVectorsSelBitset.set();
+        } else {
+            nodeGroupScanState.cachedScannedVectorsSelBitset.reset();
+            for (auto i = 0u; i < tableState.outState->getSelVector().getSelSize(); i++) {
+                nodeGroupScanState.cachedScannedVectorsSelBitset.set(
+                    tableState.outState->getSelVector()[i], true);
+            }
+        }
+    }
+}
+
+NodeGroupScanResult CSRNodeGroup::scanCommittedInMem(Transaction* transaction,
+    RelTableScanState& tableState, CSRNodeGroupScanState& nodeGroupScanState) {
+    while (true) {
+        if (tableState.currBoundNodeIdx >= tableState.cachedBoundNodeSelVector.getSelSize()) {
+            return NODE_GROUP_SCAN_EMMPTY_RESULT;
+        }
+        if (nodeGroupScanState.inMemCSRList.rowIndices.empty()) {
+            const auto boundNodePos =
+                tableState.cachedBoundNodeSelVector[tableState.currBoundNodeIdx];
+            const auto boundNodeOffset = tableState.nodeIDVector->readNodeOffset(boundNodePos);
+            const auto offsetInGroup = boundNodeOffset % StorageConstants::NODE_GROUP_SIZE;
+            nodeGroupScanState.inMemCSRList = csrIndex->indices[offsetInGroup];
+        }
+        if (!nodeGroupScanState.inMemCSRList.isSequential) {
+            KU_ASSERT(std::is_sorted(nodeGroupScanState.inMemCSRList.rowIndices.begin(),
+                nodeGroupScanState.inMemCSRList.rowIndices.end()));
+        }
+        auto scanResult =
+            nodeGroupScanState.inMemCSRList.isSequential ?
+                scanCommittedInMemSequential(transaction, tableState, nodeGroupScanState) :
+                scanCommittedInMemRandom(transaction, tableState, nodeGroupScanState);
+        if (scanResult == NODE_GROUP_SCAN_EMMPTY_RESULT) {
+            tableState.currBoundNodeIdx++;
+            nodeGroupScanState.nextRowToScan = 0;
+            nodeGroupScanState.inMemCSRList.clear();
+        } else {
+            tableState.nodeIDVector->state->getSelVectorUnsafe().setToFiltered(1);
```

Comment:
I do think we need to understand at which point we set the selected size of input nodeIDVector to 1.